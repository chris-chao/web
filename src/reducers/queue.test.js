// @flow
import {
  addItemsToQueue,
  replaceQueue,
  skipRelative,
  removeFromQueue,
  skip,
  setPlayback,
  skipToPosition,
  togglePlayback,
} from './queue';

const initialState = {
  items: [
    { id: 'a', songId: 'Legend' },
    { id: 'b', songId: 'Blessings (Explicit) ft. Drake, Kanye West' },
    { id: 'c', songId: 'Pick Me Up ft. Anna of the North' },
    { id: 'd', songId: 'No Less' },
    { id: 'e', songId: 'Love Is Gone (Audio) ft. Drew Love (of THEY.)' },
  ],
  position: 2,
  shouldBePlaying: false,
};

describe('add to queue reducer', () => {
  const items = [
    { songId: 'this is a song' },
    { songId: 'this is another song' },
  ];

  it('adds to end of queue', () => {
    const action = {
      type: 'ADD_ITEMS_TO_QUEUE',
      items,
      position: 'END',
    };
    const newState = addItemsToQueue(initialState, action);
    expect(newState).toMatchSnapshot();
  });

  it('adds after current playing item', () => {
    const action = {
      type: 'ADD_ITEMS_TO_QUEUE',
      items,
      position: 'AFTER_CURRENT',
    };
    const newState = addItemsToQueue(initialState, action);
    expect(newState).toMatchSnapshot();
  });

  it('adds before current playing item', () => {
    const action = {
      type: 'ADD_ITEMS_TO_QUEUE',
      items,
      position: 'BEFORE_CURRENT',
    };
    const newState = addItemsToQueue(initialState, action);
    expect(newState).toMatchSnapshot();
  });
});

describe('replace reducer', () => {
  it('replaces the queue', () => {
    const action = {
      type: 'REPLACE_QUEUE',
      items: [{ songId: 'Just The Way You Are' }, { songId: 'Glad You Came' }],
    };
    const newState = replaceQueue(initialState, action);
    expect(newState).toMatchSnapshot();
  });
});

describe('skip relative reducer', () => {
  it('skips forward one', () => {
    const action = {
      type: 'SKIP_RELATIVE',
      offset: 1,
    };
    const newState = skipRelative(initialState, action);
    expect(newState.position).toBe(initialState.position + 1);
  });

  it('skips backwards one', () => {
    const action = {
      type: 'SKIP_RELATIVE',
      offset: -1,
    };
    const newState = skipRelative(initialState, action);
    expect(newState.position).toBe(initialState.position - 1);
  });

  it('stays within upper bound', () => {
    const action = {
      type: 'SKIP_RELATIVE',
      offset: 10,
    };
    const newState = skipRelative(initialState, action);
    expect(newState.position).toBe(initialState.items.length);
  });
});

describe('remove reducer', () => {
  it('removes songs after the current playing song', () => {
    const action = {
      type: 'REMOVE_FROM_QUEUE',
      songs: ['d', 'e'],
    };
    const newState = removeFromQueue(initialState, action);
    expect(newState).toMatchSnapshot();
  });

  it('removes songs before the current playing song', () => {
    const action = {
      type: 'REMOVE_FROM_QUEUE',
      songs: ['a', 'b'],
    };
    const newState = removeFromQueue(initialState, action);
    expect(newState).toMatchSnapshot();
  });

  it('removes the currently playing song', () => {
    const action = {
      type: 'REMOVE_FROM_QUEUE',
      songs: ['c'],
    };
    const newState = removeFromQueue(initialState, action);
    expect(newState).toMatchSnapshot();
  });
});

describe('skip reducer', () => {
  it('skips to a song', () => {
    const action = {
      type: 'SKIP_TO',
      cursor: 'd',
    };
    const newState = skip(initialState, action);
    expect(newState.position).toBe(3);
  });

  it("doesn't skip to an invalid song", () => {
    const action = {
      type: 'SKIP_TO',
      cursor: 'invalid',
    };
    const newState = skip(initialState, action);
    expect(newState).toEqual(initialState);
  });
});

describe('skip to absolute position', () => {
  it('skips to a position', () => {
    const action = {
      type: 'SKIP_TO_POSITION',
      position: 1,
    };

    const newState = skipToPosition(initialState, action);
    expect(newState.position).toEqual(1);
  });
});

describe('set playback reducer', () => {
  it('sets playing to true', () => {
    const action = { type: 'SET_PLAYBACK', playing: true };
    const newState = setPlayback(initialState, action);
    expect(newState).toEqual({ ...initialState, shouldBePlaying: true });
  });
});

describe('toggle playback reducer', () => {
  it('changes from paused to playing', () => {
    const action = { type: 'TOGGLE_PLAYBACK' };
    const newState = togglePlayback(initialState, action);
    expect(newState.shouldBePlaying).toEqual(true);
  });

  it('changes from playing to paused', () => {
    const localState = { ...initialState, shouldBePlaying: true };
    const action = { type: 'TOGGLE_PLAYBACK' };
    const newState = togglePlayback(localState, action);
    expect(newState.shouldBePlaying).toEqual(false);
  });
});
