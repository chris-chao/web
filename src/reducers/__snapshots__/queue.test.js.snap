// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add to queue reducer adds after current playing item 1`] = `
Object {
  "items": Array [
    Object {
      "id": "a",
      "songId": "Legend",
    },
    Object {
      "id": "b",
      "songId": "Blessings (Explicit) ft. Drake, Kanye West",
    },
    Object {
      "id": "c",
      "songId": "Pick Me Up ft. Anna of the North",
    },
    Object {
      "id": "2",
      "songId": "this is a song",
    },
    Object {
      "id": "3",
      "songId": "this is another song",
    },
    Object {
      "id": "d",
      "songId": "No Less",
    },
    Object {
      "id": "e",
      "songId": "Love Is Gone (Audio) ft. Drew Love (of THEY.)",
    },
  ],
  "position": 2,
  "shouldBePlaying": false,
}
`;

exports[`add to queue reducer adds before current playing item 1`] = `
Object {
  "items": Array [
    Object {
      "id": "a",
      "songId": "Legend",
    },
    Object {
      "id": "4",
      "songId": "this is a song",
    },
    Object {
      "id": "5",
      "songId": "this is another song",
    },
    Object {
      "id": "b",
      "songId": "Blessings (Explicit) ft. Drake, Kanye West",
    },
    Object {
      "id": "c",
      "songId": "Pick Me Up ft. Anna of the North",
    },
    Object {
      "id": "d",
      "songId": "No Less",
    },
    Object {
      "id": "e",
      "songId": "Love Is Gone (Audio) ft. Drew Love (of THEY.)",
    },
  ],
  "position": 2,
  "shouldBePlaying": false,
}
`;

exports[`add to queue reducer adds to end of queue 1`] = `
Object {
  "items": Array [
    Object {
      "id": "a",
      "songId": "Legend",
    },
    Object {
      "id": "b",
      "songId": "Blessings (Explicit) ft. Drake, Kanye West",
    },
    Object {
      "id": "c",
      "songId": "Pick Me Up ft. Anna of the North",
    },
    Object {
      "id": "d",
      "songId": "No Less",
    },
    Object {
      "id": "e",
      "songId": "Love Is Gone (Audio) ft. Drew Love (of THEY.)",
    },
    Object {
      "id": "0",
      "songId": "this is a song",
    },
    Object {
      "id": "1",
      "songId": "this is another song",
    },
  ],
  "position": 2,
  "shouldBePlaying": false,
}
`;

exports[`remove reducer removes songs after the current playing song 1`] = `
Object {
  "items": Array [
    Object {
      "id": "a",
      "songId": "Legend",
    },
    Object {
      "id": "b",
      "songId": "Blessings (Explicit) ft. Drake, Kanye West",
    },
    Object {
      "id": "c",
      "songId": "Pick Me Up ft. Anna of the North",
    },
  ],
  "position": 2,
  "shouldBePlaying": false,
}
`;

exports[`remove reducer removes songs before the current playing song 1`] = `
Object {
  "items": Array [
    Object {
      "id": "c",
      "songId": "Pick Me Up ft. Anna of the North",
    },
    Object {
      "id": "d",
      "songId": "No Less",
    },
    Object {
      "id": "e",
      "songId": "Love Is Gone (Audio) ft. Drew Love (of THEY.)",
    },
  ],
  "position": 0,
  "shouldBePlaying": false,
}
`;

exports[`remove reducer removes the currently playing song 1`] = `
Object {
  "items": Array [
    Object {
      "id": "a",
      "songId": "Legend",
    },
    Object {
      "id": "b",
      "songId": "Blessings (Explicit) ft. Drake, Kanye West",
    },
    Object {
      "id": "d",
      "songId": "No Less",
    },
    Object {
      "id": "e",
      "songId": "Love Is Gone (Audio) ft. Drew Love (of THEY.)",
    },
  ],
  "position": 2,
  "shouldBePlaying": false,
}
`;

exports[`replace reducer replaces the queue 1`] = `
Object {
  "items": Array [
    Object {
      "id": "6",
      "songId": "Just The Way You Are",
    },
    Object {
      "id": "7",
      "songId": "Glad You Came",
    },
  ],
  "position": 0,
  "shouldBePlaying": false,
}
`;
